File: backend/app/schemas/saved_listings.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.listings import ListingSummary


class SavedListingCreateRequest(BaseModel):
    listing_id: str


class SavedListingItem(BaseModel):
    id: str
    listing_id: str
    saved_at: datetime
    listing_summary: ListingSummary


class SavedListingsPagination(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


class SavedListingsResponse(BaseModel):
    items: list[SavedListingItem]
    pagination: SavedListingsPagination


class SavedListingCreateResponse(BaseModel):
    id: str
    listing_id: str
    saved_at: datetime

File: frontend/src/api/savedListings.ts
import { apiClient } from "./client";
import type { ListingSummary } from "./listings";

export type SavedListingItem = {
  id: string;
  listing_id: string;
  saved_at: string;
  listing_summary: ListingSummary;
};

export type SavedListingsResponse = {
  items: SavedListingItem[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
};

export type SavedListingCreateResponse = {
  id: string;
  listing_id: string;
  saved_at: string;
};

export async function getSavedListings(page = 1, pageSize = 20): Promise<SavedListingsResponse> {
  const response = await apiClient.get<SavedListingsResponse>(`/saved-listings?page=${page}&page_size=${pageSize}`);
  return response.data;
}

export async function createSavedListing(listingId: string): Promise<SavedListingCreateResponse> {
  const response = await apiClient.post<SavedListingCreateResponse>("/saved-listings", { listing_id: listingId });
  return response.data;
}

export async function deleteSavedListing(savedListingId: string): Promise<void> {
  await apiClient.delete(`/saved-listings/${savedListingId}`);
}
